File: plateform/frontend/src/types/networkHealth.ts
export type ReachabilityStatus =
  | "reachable"
  | "unreachable"
  | "unknown"
  | "not_supported"
  | "not_checked";

export type SourceMode = "seeded_demo" | "nokia_simulator" | "nokia_live";

export interface NetworkFilters {
  search: string;
  zone: string;
  asset_type: string;
  reachability: string;
  location_available: string;
  cellular_available: string;
  source_mode: string;
  device: string;
}

export interface LastTelemetry {
  observed_at: string | null;
  pressure_kpa: number | null;
  flow_lps: number | null;
  temperature_c: number | null;
}

export interface DeviceNetworkSnapshot {
  snapshot_id: string;
  asset_id: string;
  asset_name: string;
  asset_type: string;
  zone_id: string | null;
  zone_name: string | null;
  location_label: string | null;
  registered_latitude: number | null;
  registered_longitude: number | null;
  has_cellular_identity: boolean;
  device_msisdn_masked: string | null;
  reachability_status: ReachabilityStatus;
  reachable_via: string | null;
  reachability_checked_at: string | null;
  network_location_available: boolean;
  network_latitude: number | null;
  network_longitude: number | null;
  accuracy_radius_m: number | null;
  location_area_type: string | null;
  location_observed_at: string | null;
  location_offset_m: number | null;
  retrieved_at: string | null;
  stale: boolean;
  provider: string | null;
  source_mode: SourceMode | null;
  data_source_label: string;
  error_code: string | null;
  error_message: string | null;
  last_telemetry: LastTelemetry | null;
}

export interface DeviceNetworkHistoryItem {
  snapshot_id: string;
  reachability_status: ReachabilityStatus;
  reachable_via: string | null;
  network_location_available: boolean;
  network_latitude: number | null;
  network_longitude: number | null;
  accuracy_radius_m: number | null;
  retrieved_at: string;
  source_mode: SourceMode;
  error_code: string | null;
}

export interface DeviceNetworkDetail extends DeviceNetworkSnapshot {
  snapshot_public_id: string | null;
  request_correlation_id: string | null;
  cached: boolean;
  history: DeviceNetworkHistoryItem[];
}

export interface DeviceNetworkListResponse {
  items: DeviceNetworkSnapshot[];
  total: number;
  source_mode: SourceMode;
  data_source_label: string;
  last_refresh_at: string | null;
}

export interface DeviceNetworkSummary {
  cellular_devices: number;
  reachable: number;
  unreachable: number;
  unknown_or_not_checked: number;
  network_location_available: number;
  stale_checks: number;
  last_refresh_at: string | null;
  source_mode: SourceMode;
  data_source_label: string;
  environment_label: string;
  available_zones: string[];
  note: string;
}

export interface BulkRefreshResponse {
  requested: number;
  succeeded: number;
  failed: number;
  skipped_cached: number;
  items: DeviceNetworkSnapshot[];
  errors: { asset_id: string; error_code: string; error_message: string }[];
  source_mode: SourceMode;
  data_source_label: string;
  partial_success: boolean;
}

export interface IncidentDeviceNetworkContext {
  incident_id: string;
  asset_id: string | null;
  available: boolean;
  reachability_status: ReachabilityStatus | null;
  reachable_via: string | null;
  reachability_checked_at: string | null;
  network_location_available: boolean;
  network_latitude: number | null;
  network_longitude: number | null;
  accuracy_radius_m: number | null;
  registered_latitude: number | null;
  registered_longitude: number | null;
  location_offset_m: number | null;
  source_mode: SourceMode | null;
  data_source_label: string | null;
  device_msisdn_masked: string | null;
  note: string;
}

File: plateform/backend/app/schemas/network_health.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.network.constants import MOCK_DATA_MODE

ReachabilityStatus = Literal["reachable", "unreachable", "unknown", "not_supported", "not_checked"]
SourceMode = Literal["seeded_demo", "nokia_simulator", "nokia_live"]


class NetworkEventSummary(BaseModel):
    id: str
    public_id: str
    event_id: str
    event_type: str
    status: str
    summary: str
    cluster_id: str | None
    device_id: str | None
    incident_id: str | None
    occurred_at: datetime
    data_mode: str = MOCK_DATA_MODE
    contract_version: str


class NetworkEventDetail(NetworkEventSummary):
    payload: dict[str, Any]
    input_summary: dict[str, Any]
    output_summary: dict[str, Any]
    error_code: str | None
    error_message: str | None


class NetworkEventListResponse(BaseModel):
    items: list[NetworkEventSummary]
    total: int
    page: int
    page_size: int
    data_mode: str = MOCK_DATA_MODE
    note: str = "Historical Network Agent audit records — not current device health."


class LastTelemetry(BaseModel):
    observed_at: datetime | None = None
    pressure_kpa: float | None = None
    flow_lps: float | None = None
    temperature_c: float | None = None


class DeviceNetworkSnapshotSummary(BaseModel):
    snapshot_id: str
    asset_id: str
    asset_name: str
    asset_type: str
    zone_id: str | None
    zone_name: str | None
    location_label: str | None
    registered_latitude: float | None
    registered_longitude: float | None
    has_cellular_identity: bool
    device_msisdn_masked: str | None
    reachability_status: ReachabilityStatus
    reachable_via: str | None
    reachability_checked_at: datetime | None
    network_location_available: bool
    network_latitude: float | None
    network_longitude: float | None
    accuracy_radius_m: float | None
    location_area_type: str | None
    location_observed_at: datetime | None
    location_offset_m: float | None
    retrieved_at: datetime | None
    stale: bool
    provider: str | None
    source_mode: SourceMode | None
    data_source_label: str
    error_code: str | None = None
    error_message: str | None = None
    last_telemetry: LastTelemetry | None = None


class DeviceNetworkSnapshotDetail(DeviceNetworkSnapshotSummary):
    snapshot_public_id: str | None = None
    request_correlation_id: str | None = None
    cached: bool = False
    history: list["DeviceNetworkHistoryItem"] = Field(default_factory=list)


class DeviceNetworkHistoryItem(BaseModel):
    snapshot_id: str
    reachability_status: ReachabilityStatus
    reachable_via: str | None
    network_location_available: bool
    network_latitude: float | None
    network_longitude: float | None
    accuracy_radius_m: float | None
    retrieved_at: datetime
    source_mode: SourceMode
    error_code: str | None


class DeviceNetworkListResponse(BaseModel):
    items: list[DeviceNetworkSnapshotSummary]
    total: int
    source_mode: SourceMode
    data_source_label: str
    last_refresh_at: datetime | None


class DeviceNetworkSummary(BaseModel):
    cellular_devices: int
    reachable: int
    unreachable: int
    unknown_or_not_checked: int
    network_location_available: int
    stale_checks: int
    last_refresh_at: datetime | None
    source_mode: SourceMode
    data_source_label: str
    environment_label: str
    available_zones: list[str]
    note: str = "Device network observations from AquaPulse snapshots. Nokia reachability is not incident status."


class DeviceNetworkRefreshRequest(BaseModel):
    force: bool = False


class BulkDeviceNetworkRefreshRequest(BaseModel):
    force: bool = False
    asset_ids: list[str] | None = None
    limit: int = Field(default=20, ge=1, le=50)


class BulkRefreshError(BaseModel):
    asset_id: str
    error_code: str
    error_message: str


class BulkDeviceNetworkRefreshResponse(BaseModel):
    requested: int
    succeeded: int
    failed: int
    skipped_cached: int
    items: list[DeviceNetworkSnapshotSummary]
    errors: list[BulkRefreshError]
    source_mode: SourceMode
    data_source_label: str
    partial_success: bool


class IncidentDeviceNetworkContext(BaseModel):
    incident_id: str
    asset_id: str | None
    available: bool
    reachability_status: ReachabilityStatus | None = None
    reachable_via: str | None = None
    reachability_checked_at: datetime | None = None
    network_location_available: bool = False
    network_latitude: float | None = None
    network_longitude: float | None = None
    accuracy_radius_m: float | None = None
    registered_latitude: float | None = None
    registered_longitude: float | None = None
    location_offset_m: float | None = None
    source_mode: SourceMode | None = None
    data_source_label: str | None = None
    device_msisdn_masked: str | None = None
    note: str = (
        "Nokia reachability is not incident status. "
        "Investigating, waiting for approval, and resolved remain incident workflow states."
    )


DeviceNetworkSnapshotDetail.model_rebuild()
